packages/arb-util/core/execution.go

```diff
@@ -145,8 +145,15 @@ func IsAssertionValid(assertion *Assertion, execTracker *ExecutionTracker, targe
 	if err != nil {
 		return false, err
 	}
-	notEnoughMessages := localExecutionInfo.InboxMessagesRead().Cmp(assertion.InboxMessagesRead()) < 0
-	if notEnoughMessages {
+	if localExecutionInfo.InboxMessagesRead().Cmp(assertion.InboxMessagesRead()) < 0 {
+		// We didn't read enough messages.
+		// This can either mean that our messages lasted longer, or that we are missing messages.
+		if localExecutionInfo.After.TotalGasConsumed.Cmp(assertion.After.TotalGasConsumed) < 0 {
+			// This means we stopped because we're missing messages,
+			// but the on-chain rollup must've had these messages.
+			// Error and try again when we have the messages.
+			return false, errors.New("Missing messages to evaluate assertion")
```

Comment:
I'd appreciate if someone thought through this logic and made sure it's sound. I think it is but I might've missed an edge case.

Reply:
I think the other possibility here is that we're in at stopped or errored state